### PR TITLE
GH #69: Redesigned dir() (and some file()) interface

### PR DESCRIPTION
**Pull request by xsawyerx**
There are some significant changes here, but they essentially boil down to:

* You cannot set directory content (not through `->dir()` or through `->contents()`)

```
$d = Test::MockFile->dir( $name, [@content] ); # dies
$d->contents(...);                             # dies
```

* You cannot set stats for either a directory without using the right Perl core keywords. Not everything is supported yet, but it is now a priority to fill in these. (A helper method might be added in the future as well.)

```
# This fails
$d = Test::MockFile->dir( $name, [], { 'mode' => 0755 } ); # dies

# This works
$d = Test::MockFile->dir($name);
mkdir $dirname, 0755;
```

* To create a directory, you can use `mkdir()` or you can create files in it using `->file()`. This means that if you create files with `->file()` and then call `->dir()`, that directory will exist already.

```
# Cannot create directory content through dir()
$f = Test::MockFile->file( $name, undef, \%stats ); # dies
$d = Test::MockFile->dir( $name, [], \%stats );     # dies

# Creating content with ->file() works
$f = Test::MockFile->file( '/dir/foo' => 'hello' );
$d = Test::MockFile->dir( '/dir' );
-d '/dir';                                          # success
$d->contnets();                                     # [ . .. foo ]

# Creating a directory with mkdir() works
$d = Test::MockFile->dir( '/dir' );
-d '/dir';                              # fails because it doesn't exist yet
mkdir '/dir';                           # success because you created it
-d '/dir';                              # success because it now exists
$d->contents();                         # [ . .. ]

# However, an empty file does not create a directory
$f = Test::MockFile->file('/dir/foo');
$d = Test::MockFile->dir( '/dir' );
-d '/dir';                              # fails
```

* This also means that if you created a nonexistent file, create a nonexistent directory, and then give the file content, the directory will automatically be created by it.

```
$f = Test::MockFile->file('/dir/foo');
$d = Test::MockFile->dir( '/dir' );
-d '/dir';                              # fails, not created yet
$f->contents('hello');                  # Create file with a value
-d '/dir';                              # success because we created the file
```

* There was an edge-case in which a symlink'ed file can turn into a directory with `mkdir()`. This is not supported anymore.

```
my $file = Test::MockFile->file( '/foo/a', 'hello' );
$file->symlink( '/foo/a_link' => $file );
$file->unlink();
mkdir('/foo/a');                                      # fails

# Instead, undef it (for cleanup) and create a new object
undef $file;
$file = Test::MockFile->dir('/foo/a');
mkdir '/foo/a';                                       # success

# Or use a different scope for automatic cleanup
{
    my $file = Test::MockFile->file( '/foo/a', 'hello' );
    $file->symlink( '/foo/a_link' => $file );
}

{
    $file = Test::MockFile->dir('/foo/a');
    mkdir '/foo/a';
}
```

Implementation details:

* We introduce `has_content` as a key that tells us the directory does exist, even if it
  has no files in it and that the value has to be determined at runtime.

* We try to use `->contents()` everywhere so we generate the content for directories based on the existing files we already mocked with values.

* Calling either `mkdir` or creating a new file with content in a directory using `->file()` will both set `has_content` to tell us this directory now exists.

* mkdir() had an edge-case to turn an unlinked symlink mocked object into a directory
  * This isn't supported anymore: `undef` it and start over like a normal person

Resolved GH #69.

### CODE REVIEW

File: t/dir_interface.t [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+});
+
+subtest( 'Scenario 4: Creating ->file() with content creates dir' => sub {
+    my $dirname = $get_dirname->();
+    my $dir     = Test::MockFile->dir($dirname);
+
+    ok( !-d $dirname, "$dirname does not exist yet" );
+    my $file = Test::MockFile->file( "$dirname/foo", 'some content' );
+    ok( -d $dirname, "$dirname now exists, after creating file with content" );
+    ok( !mkdir($dirname), "mkdir $dirname fails, since dir already exists" );
```
Comment (toddr) [inline t/dir_interface.t:130]: It feels like this should fail because it's not mocked.
